File: bci/api/python/wia_bci/adapters/base_adapter.py
"""WIA BCI Base Adapter."""

from abc import ABC, abstractmethod
from typing import Callable, Optional

from wia_bci.types.config import DeviceConfig, DeviceType
from wia_bci.types.device import ChannelInfo, DeviceInfo
from wia_bci.types.events import SignalEvent, ErrorEvent


class BaseBciAdapter(ABC):
    """
    Base BCI Adapter Interface.

    All device-specific adapters must inherit from this class.
    """

    def __init__(self) -> None:
        self._connected = False
        self._streaming = False
        self._device_info: Optional[DeviceInfo] = None
        self._channels: list[ChannelInfo] = []
        self._data_handler: Optional[Callable[[SignalEvent], None]] = None
        self._error_handler: Optional[Callable[[ErrorEvent], None]] = None

    @property
    @abstractmethod
    def type(self) -> DeviceType:
        """Device type this adapter handles."""
        ...

    @property
    @abstractmethod
    def name(self) -> str:
        """Adapter name."""
        ...

    @abstractmethod
    async def connect(self, config: DeviceConfig) -> None:
        """Connect to the BCI device."""
        ...

    @abstractmethod
    async def disconnect(self) -> None:
        """Disconnect from the BCI device."""
        ...

    @abstractmethod
    async def start_stream(self) -> None:
        """Start data streaming."""
        ...

    @abstractmethod
    async def stop_stream(self) -> None:
        """Stop data streaming."""
        ...

    def is_connected(self) -> bool:
        """Check if connected."""
        return self._connected

    def is_streaming(self) -> bool:
        """Check if streaming."""
        return self._streaming

    def get_device_info(self) -> Optional[DeviceInfo]:
        """Get device information."""
        return self._device_info

    def get_channels(self) -> list[ChannelInfo]:
        """Get channel information."""
        return self._channels

    def on_data(self, handler: Callable[[SignalEvent], None]) -> None:
        """Register data handler."""
        self._data_handler = handler

    def on_error(self, handler: Callable[[ErrorEvent], None]) -> None:
        """Register error handler."""
        self._error_handler = handler

    def dispose(self) -> None:
        """Clean up resources."""
        self._data_handler = None
        self._error_handler = None

    def _emit_data(self, data: SignalEvent) -> None:
        """Emit data event to handler."""
        if self._data_handler:
            self._data_handler(data)

    def _emit_error(self, error: ErrorEvent) -> None:
        """Emit error event to handler."""
        if self._error_handler:
            self._error_handler(error)

File: bci/api/typescript/src/adapters/BaseAdapter.ts
/**
 * WIA BCI Base Adapter Interface
 * @module wia-bci/adapters/BaseAdapter
 */

import type {
  DeviceType,
  DeviceConfig,
  DeviceInfo,
  ChannelInfo,
  SignalEvent,
  ErrorEvent,
} from '../types';

/**
 * BCI Adapter Interface
 * All device-specific adapters must implement this interface
 */
export interface IBciAdapter {
  /** Device type this adapter handles */
  readonly type: DeviceType;
  /** Adapter name */
  readonly name: string;

  /**
   * Connect to the BCI device
   * @param config Device configuration
   */
  connect(config: DeviceConfig): Promise<void>;

  /**
   * Disconnect from the BCI device
   */
  disconnect(): Promise<void>;

  /**
   * Check if connected
   */
  isConnected(): boolean;

  /**
   * Start data streaming
   */
  startStream(): Promise<void>;

  /**
   * Stop data streaming
   */
  stopStream(): Promise<void>;

  /**
   * Check if streaming
   */
  isStreaming(): boolean;

  /**
   * Get device information
   */
  getDeviceInfo(): DeviceInfo | null;

  /**
   * Get channel information
   */
  getChannels(): ChannelInfo[];

  /**
   * Register data handler
   * @param handler Function to call when data is received
   */
  onData(handler: (data: SignalEvent) => void): void;

  /**
   * Register error handler
   * @param handler Function to call when error occurs
   */
  onError(handler: (error: ErrorEvent) => void): void;

  /**
   * Clean up resources
   */
  dispose(): void;
}

/**
 * Abstract Base Adapter
 * Provides common functionality for all adapters
 */
export abstract class BaseAdapter implements IBciAdapter {
  abstract readonly type: DeviceType;
  abstract readonly name: string;

  protected _connected = false;
  protected _streaming = false;
  protected deviceInfo: DeviceInfo | null = null;
  protected channels: ChannelInfo[] = [];
  protected dataHandler?: (data: SignalEvent) => void;
  protected errorHandler?: (error: ErrorEvent) => void;

  abstract connect(config: DeviceConfig): Promise<void>;
  abstract disconnect(): Promise<void>;
  abstract startStream(): Promise<void>;
  abstract stopStream(): Promise<void>;

  isConnected(): boolean {
    return this._connected;
  }

  isStreaming(): boolean {
    return this._streaming;
  }

  getDeviceInfo(): DeviceInfo | null {
    return this.deviceInfo;
  }

  getChannels(): ChannelInfo[] {
    return this.channels;
  }

  onData(handler: (data: SignalEvent) => void): void {
    this.dataHandler = handler;
  }

  onError(handler: (error: ErrorEvent) => void): void {
    this.errorHandler = handler;
  }

  dispose(): void {
    if (this._streaming) {
      this.stopStream().catch(() => {});
    }
    if (this._connected) {
      this.disconnect().catch(() => {});
    }
    this.dataHandler = undefined;
    this.errorHandler = undefined;
  }

  /**
   * Emit data event to handler
   */
  protected emitData(data: SignalEvent): void {
    this.dataHandler?.(data);
  }

  /**
   * Emit error event to handler
   */
  protected emitError(error: ErrorEvent): void {
    this.errorHandler?.(error);
  }
}
